File: packages/native-shell-ios/Sources/SelfNativeShell/WebView/RemoteContentIntegrity.swift
// SPDX-License-Identifier: BUSL-1.1

import Foundation

// nil mimeType is allowed — some CDNs omit Content-Type; the SHA-256 hash is the primary integrity gate.
enum RemoteContentIntegrity {
    static func normalizeSha256(_ value: String) -> String {
        var normalized = value.lowercased()
        if normalized.hasPrefix("sha256-") {
            normalized.removeFirst("sha256-".count)
        }
        if normalized.hasPrefix("0x") {
            normalized.removeFirst(2)
        }
        return normalized
    }

    static func isAcceptableMimeType(_ mimeType: String?) -> Bool {
        let normalized = mimeType?
            .split(separator: ";", maxSplits: 1)
            .first?
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .lowercased()
        return normalized == nil || normalized == "text/html"
    }
}

File: packages/native-shell-android/src/main/kotlin/xyz/self/sdk/webview/RemoteContentIntegrity.kt
// SPDX-License-Identifier: BUSL-1.1

package xyz.self.sdk.webview

import java.security.MessageDigest

// null contentType is allowed — some CDNs omit Content-Type; the SHA-256 hash is the primary integrity gate.
internal object RemoteContentIntegrity {
    fun normalizeSha256(value: String): String = value.lowercase().removePrefix("sha256-").removePrefix("0x")

    fun sha256Hex(bytes: ByteArray): String =
        MessageDigest.getInstance("SHA-256").digest(bytes).joinToString("") { byte ->
            "%02x".format(byte)
        }

    fun isAcceptableContentType(rawContentType: String?): Boolean {
        val normalized = rawContentType?.substringBefore(";")?.trim()?.lowercase()
        return normalized == null || normalized == "text/html"
    }
}
